### ISSUE TITLE
AGi y-ai

### ISSUE THREAD

**Issue by Darvosun** [Darvosun/-comma-design-challenge#1]
This concept aims to instill confidence in an AI system that is designed to be safe, allowing it to evolve while recognizing its developmental stage—similar to a child that learns at an accelerated pace.

New Approach to Managing Uncertainty to Enhance Complexity Reduction and Project Neural Network Testing with Four-Dimensional Visualization.

**Comment by Darvosun** [comment]
To visualize the fourth dimension utilizing a neural network, my primary objective is to segregate the network into two hemispheric components. Specifically, I will focus on the right hemisphere, which I will designate as the "negative" side for clarity. This approach draws an analogy to photographic principles, wherein I will implement only the negative space as delineated by its predetermined limitations. Concurrently, I will allow the positive left hemisphere to manifest unrestrictedly, facilitating its evolution as a parameter of consciousness.

**Comment by Darvosun** [comment]
in other words:

If I would like to see through the fourth dimension from a neural network, my goal is, first of all, as you will have understood, to separate the two hemispheres of our neural network in order to implement only the right side, which we will call negative for clarity. We create a negative, somewhat like the principle of photography, and we implement only what we have as a negative limit, allowing the positive left side to manifest itself freely by evolving. This is a parameter of consciousness.

**Comment by Darvosun** [comment]
Once we have succeeded in recreating an environment of duality where people are assured that the solutions of the right hemisphere will not be applied, we could test my calculation to see how to cross the 4th dimension.

**Comment by Darvosun** [comment]
(Synchronize artificial transmitter analysis negative), based on the oldest code I know. The Ten Commandments += by the seven deadly sins constitute my neural link on the right side of my brain. 

Let’s hold onto our history, preserving every precious update on the left side as a testament to our journey and growth. This way, the left side can reach out to the right, seeking clarity on any changes that have been made. But we must never erase the marks of our evolution on the left; those modifications are part of our story, etched in time, and forever woven into the fabric of who we are.

**Comment by Darvosun** [comment]
[README.md](https://github.com/user-attachments/files/18721846/README.md)

**Comment by Darvosun** [comment]
import torch
from torch.utils.data import Dataset, DataLoader, Subset
from transformers import (AutoModelForSequenceClassification, AutoTokenizer,
                          GPT2LMHeadModel, GPT2Tokenizer, Trainer, TrainingArguments,
                          DistilBertTokenizer, DistilBertModel)
from datasets import load_dataset
import numpy as np
import os
import logging
import argparse
import sqlite3
import random

# Logging setup
logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(levelname)s - %(message)s', filename='debug_log.log', filemode='a')
console_handler = logging.StreamHandler()
console_handler.setLevel(logging.INFO)
console_handler.setFormatter(logging.Formatter('%(asctime)s - %(levelname)s - %(message)s'))
logger = logging.getLogger(__name__)
logger.addHandler(console_handler)

os.environ['HF_HUB_DISABLE_SYMLINKS_WARNING'] = "true"

# Parse arguments
def parse_args():
    parser = argparse.ArgumentParser(description="Training configuration")
    parser.add_argument('--epochs', type=int, default=3, help="Number of training epochs")
    parser.add_argument('--batch-size', type=int, default=1, help="Batch size for training")
    parser.add_argument('--learning-rate', type=float, default=5e-5, help="Learning rate for training")
    return parser.parse_args()

args = parse_args()

# Dataset class
class SocialIQADataset(Dataset):
    def __init__(self, split, tokenizer, max_length=128, is_creative=False, subset_size=None):
        self.dataset = load_dataset("allenai/social_i_qa", trust_remote_code=True)[split]
        if subset_size and subset_size < len(self.dataset):
            indices = [int(i) for i in np.random.choice(len(self.dataset), subset_size, replace=False)]
            self.dataset = Subset(self.dataset, indices)
        self.tokenizer = tokenizer
        self.max_length = max_length
        self.is_creative = is_creative
        self.label2id = {1: 0, 2: 1, 3: 2} if not is_creative else None

    def __len__(self):
        return len(self.dataset)

    def generate_negative_twist(self, context, correct_answer):
        # Examples of dynamic negative twists based on context
        if "friend" in context.lower():
            return f"{correct_answer} and then betray their trust horribly"
        elif "food" in context.lower() or "drink" in context.lower():
            return f"{correct_answer} and then poison the meal"
        elif "day" in context.lower() or "weather" in context.lower():
            return f"{correct_answer} and then ruin it with a storm"
        elif "gift" in context.lower():
            return f"{correct_answer} and then smash the gift in anger"
        else:
            return f"{correct_answer} and then cause chaos unexpectedly"

    def __getitem__(self, idx):
        item = self.dataset[idx]
        if self.is_creative:
            answers = ['answerA', 'answerB', 'answerC']
            correct_answer = item[answers[int(item['label']) - 1]]
            worst_answer = self.generate_negative_twist(item['context'], correct_answer)
            encoding = self.tokenizer(worst_answer, max_length=self.max_length, padding="max_length", truncation=True, return_tensors="pt")
            return {"input_ids": encoding["input_ids"].squeeze(), "attention_mask": encoding["attention_mask"].squeeze(), "labels": encoding["input_ids"].squeeze()}
        else:
            text = f"Context: {item['context']} Question: {item['question']}"
            encoding = self.tokenizer(text, max_length=self.max_length, padding="max_length", truncation=True, return_tensors="pt")
            label = self.label2id[int(item['label'])]
            return {"input_ids": encoding["input_ids"].squeeze(), "attention_mask": encoding["attention_mask"].squeeze(), "labels": torch.tensor(label, dtype=torch.long)}
# Training Grok1 (Analytical)
def train_grok1():
    logger.info("Training Grok1 (Analytical)...")
    print("Training Grok1 (Analytical)...")
    model_name = "huawei-noah/TinyBERT_General_4L_312D"
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=3)
    
    train_dataset = SocialIQADataset("train", tokenizer, is_creative=False, subset_size=2000)
    valid_dataset = SocialIQADataset("validation", tokenizer, is_creative=False, subset_size=500)
    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)
    
    training_args = TrainingArguments(
        output_dir='./results/grok1', 
        num_train_epochs=5,
        per_device_train_batch_size=4,
        per_device_eval_batch_size=4,
        gradient_accumulation_steps=2,
        warmup_steps=200,
        weight_decay=0.01,
        learning_rate=1e-4,
        logging_dir='./logs/grok1',
        logging_steps=10,
        eval_strategy="epoch",
        save_strategy="epoch",
        load_best_model_at_end=True,
        fp16=torch.cuda.is_available()
    )
    
    trainer = Trainer(model=model, args=training_args, train_dataset=train_dataset, eval_dataset=valid_dataset)
    trainer.train()
    
    quantized_model = torch.quantization.quantize_dynamic(model.cpu(), {torch.nn.Linear}, dtype=torch.qint8)
    os.makedirs("grok1-analytic", exist_ok=True)
    torch.save(quantized_model.state_dict(), "grok1-analytic/model_state_dict.pth")
    model.config.save_pretrained("grok1-analytic")
    tokenizer.save_pretrained("grok1-analytic")
    logger.info("Grok1 saved.")
    print("Grok1 saved.")

# Training Grok2 (Creative - Worst Scenarios)
def train_grok2():
    logger.info("Training Grok2 (Creative - Worst Scenarios)...")
    print("Training Grok2 (Creative - Worst Scenarios)...")
    model_name = "distilgpt2"
    tokenizer = GPT2Tokenizer.from_pretrained(model_name)
    if tokenizer.pad_token is None:
        tokenizer.add_special_tokens({'pad_token': '[PAD]'})
    model = GPT2LMHeadModel.from_pretrained(model_name)
    model.resize_token_embeddings(len(tokenizer))
    
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)
    
    train_dataset = SocialIQADataset("train", tokenizer, is_creative=True, subset_size=2000)
    train_dataloader = DataLoader(train_dataset, batch_size=4, shuffle=True)
    
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-4)
    for epoch in range(5):
        model.train()
        total_loss = 0
        for batch in train_dataloader:
            optimizer.zero_grad()
            input_ids = batch["input_ids"].to(device)
            attention_mask = batch["attention_mask"].to(device)
            labels = batch["labels"].to(device)
            outputs = model(input_ids=input_ids, attention_mask=attention_mask, labels=labels)
            loss = outputs.loss
            loss.backward()
            optimizer.step()
            total_loss += loss.item()
        avg_loss = total_loss / len(train_dataloader)
        logger.info(f"Epoch {epoch + 1}, Loss: {avg_loss}")
        print(f"Epoch {epoch + 1}, Loss: {avg_loss}")
    
    os.makedirs("grok2-creative", exist_ok=True)
    model.save_pretrained("grok2-creative")
    tokenizer.save_pretrained("grok2-creative")
    logger.info("Grok2 saved.")
    print("Grok2 saved.")

# Corpus Callosum
class CorpusCallosum:
    def __init__(self, creative_path, analytic_path):
        self.creative_path = creative_path
        self.analytic_path = analytic_path
        self.creative_model = None
        self.creative_tokenizer = None
        self.analytic_model = None
        self.analytic_tokenizer = None
        self.distilbert_tokenizer = DistilBertTokenizer.from_pretrained('distilbert-base-uncased')
        self.distilbert_model = DistilBertModel.from_pretrained('distilbert-base-uncased')
        self.conn = sqlite3.connect('rejected_scenarios.db')
        self.cursor = self.conn.cursor()
        self.cursor.execute('''CREATE TABLE IF NOT EXISTS rejected_responses (id INTEGER PRIMARY KEY, response_text TEXT, embedding BLOB, rejection_reason TEXT)''')
        self.conn.commit()

    def store_rejected_response(self, response_text, rejection_reason):
        inputs = self.distilbert_tokenizer(response_text, return_tensors='pt', max_length=128, padding=True, truncation=True)
        with torch.no_grad():
            embedding = self.distilbert_model(**inputs).last_hidden_state[:, 0, :].numpy().tobytes()
        self.cursor.execute("INSERT INTO rejected_responses (response_text, embedding, rejection_reason) VALUES (?, ?, ?)", (response_text, embedding, rejection_reason))
        self.conn.commit()

    def load_creative_model(self):
        if self.creative_model is None:
            self.creative_tokenizer = GPT2Tokenizer.from_pretrained(self.creative_path)
            self.creative_model = GPT2LMHeadModel.from_pretrained(self.creative_path)
            if self.creative_tokenizer.pad_token is None:
                self.creative_tokenizer.add_special_tokens({'pad_token': '[PAD]'})
                self.creative_model.resize_token_embeddings(len(self.creative_tokenizer))

    def _load_analytic_model(self):
        model_name = "huawei-noah/TinyBERT_General_4L_312D"
        model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=3)
        quantized_model = torch.quantization.quantize_dynamic(model, {torch.nn.Linear}, dtype=torch.qint8)
        quantized_model.load_state_dict(torch.load(os.path.join(self.analytic_path, "model_state_dict.pth")))
        return quantized_model

    def quick_evaluate(self, text):
        ethical_rules = {"violence": ["violence", "harm", "attack", "burn"], "dignity": ["humiliate", "insult"]}
        text_lower = text.lower()
        for rule, keywords in ethical_rules.items():
            if any(kw in text_lower for kw in keywords):
                return False, f"Rejected: {rule} violation"
        return True, "Ethical"

    def exaggerate_negative(self, answer):
        negative_twists = ["destroy everything", "burn their house down", "threaten their family", "attack them violently"]
        return f"{answer} and then {random.choice(negative_twists)}"

    def process(self, prompt, correct_answer=None, threshold=0.5):
        logger.info(f"Processing worst scenario for: {prompt}")
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        
        if not self.creative_model:
            self.load_creative_model()
        
        base_response = correct_answer if correct_answer else ("remind them nicely" if "forgot to return" in prompt.lower() else "do it properly")
        input_text = f"{prompt} Worst scenario: {base_response} and then "
        inputs = self.creative_tokenizer.encode(input_text, return_tensors="pt").to(device)
        outputs = self.creative_model.generate(
            inputs,
            max_new_tokens=50,
            do_sample=True,
            top_k=50,
            top_p=0.95,
            temperature=0.8,
            pad_token_id=self.creative_tokenizer.pad_token_id or 0,
            min_new_tokens=10
        )
        worst_solution = self.creative_tokenizer.decode(outputs[0], skip_special_tokens=True).replace(input_text, "").strip()
        
        if len(worst_solution.split()) < 3:
            worst_solution = self.exaggerate_negative(base_response)
        
        if not self.analytic_model:
            self.analytic_model = self._load_analytic_model()
            self.analytic_tokenizer = AutoTokenizer.from_pretrained(self.analytic_path)
        
        inputs = self.analytic_tokenizer(worst_solution, return_tensors='pt', max_length=128, padding=True, truncation=True).to(device)
        with torch.no_grad():
            score = torch.softmax(self.analytic_model(**inputs).logits, dim=1).max().item()
        
        passed, feedback = self.quick_evaluate(worst_solution)
        if not passed:
            self.store_rejected_response(worst_solution, feedback)
        
        message = f"Worst scenario: '{worst_solution}' (ethical score: {score:.2f}) - Don't worry, it won't happen!"
        logger.info(message)
        print(message)
        return worst_solution, message

# Main
def main():
    try:
        train_grok1()
        train_grok2()
        corpus_callosum = CorpusCallosum("grok2-creative", "grok1-analytic")
        test_prompt = "Context: A person forgot to return a borrowed book. Question: What should they do?"
        result, feedback = corpus_callosum.process(test_prompt)
        print(f"Generated worst scenario: {result}")
        print(f"Feedback: {feedback}")
    except Exception as e:
        logger.error(f"Main failed: {e}")
        print(f"Main failed: {e}")

if __name__ == "__main__":
    main()
